### PR TITLE
Handle upload of miniscreen projects

### PR DESCRIPTION
**Pull request by jcapona**
| Status  | Ticket/Issue |
| :---: | :--: |
| Ready/Hold | [Ticket](https://pi-top.atlassian.net/browse/OS-1272) |

#### Main changes
- Support challenges that have a miniscreen project in them. Files will be copied to `~/Desktop/Projects` where the miniscreen app will find them.

#### Screenshots (feature, test output, profiling, dev tools etc)

[insert screenshots here]

#### Other notes (e.g. implementation quirks, edge cases, questions / issues)
-

#### Manual testing tips
-

#### Tag anyone who definitely needs to review or help
-

### CODE REVIEW

File: further_link/endpoint/upload.py [hunk 1]
```diff
@@ -18,8 +26,17 @@ async def upload(request):
...
 
+        # Upload to further directory
         await do_upload(directory, work_dir, user)
 
+        # Upload to projects directory to run using miniscreen app
+        if is_miniscreen_project(directory):
+            if user is None:
+                user = default_user()
+            user_home = path.expanduser(f"~{user}")
+            project_work_dir = f"{user_home}/Desktop/Projects"
```
Comment (angusjfw) [inline further_link/endpoint/upload.py:37]: Also, the location is meant to be {user_home}/Desktop/Projects/**Further**. I think this is important so that we can keep them separate in the miniscreen menu, implement the 'delete all' option for them, etc.

Actually, to be honest at some point the plan was to store the projects separately for each Further username and I'm not sure now why that didn't make it into the final spec. Maybe we should rethink it with that feature as it's important for shared school devices, although changes will be needed on Further frontend to provide the username in that case...

File: further_link/endpoint/upload.py [hunk 2]
```diff
@@ -18,8 +26,17 @@ async def upload(request):
...
+            await do_upload(directory, project_work_dir, user, use_cache=False)
```
Comment (angusjfw) [inline further_link/endpoint/upload.py:38]: the ticket was somewhat unclear about the caching/symlinking. I imagined that this would be implemented by copying the directory written above to the new location, and so when I said no cache symlink I just meant that it would be better to have an actual copy of the files there not symlinks. Now that I think about it more I can't think what would be wrong with having symlinks anyway, although I also don't see an advantage (well, maybe that the copy would be faster)

Anyway, it would be good to avoid redownloading the cacheable files each time. If you let use_cache be true it will do that by creating a second cache of the cacheable files in the project_work_dir. I think actually what we should do is not use do_upload at all here, and actually just copy the directory created by the first do_upload to the new location.

**Review by angusjfw — CHANGES_REQUESTED** [review]
I've suggested a couple of refactors: creating a helper for the projects directory location and copying the directory uploaded for further rather than re-writing/downloading the files with do_upload (possibly with deference of symlinks like `cp -L`, since the further upload location might be made temporary in future).

Also, the location the files are going to is not right according to the ticket, but actually I think we should discuss that some more since I can't remember why we didn't go for username-separated folders of these and I think that would be better. Sorry about that.

File: further_link/endpoint/upload.py [hunk 3]
```diff
@@ -1,9 +1,17 @@
...
 
 from ..util.upload import BadUpload, directory_is_valid, do_upload
-from ..util.user_config import get_working_directory
+from ..util.user_config import default_user, get_working_directory
+
+
+def is_miniscreen_project(directory):
+    for filename in directory.get("files", {}):
+        if filename.endswith(".cfg"):
+            return True
```
[the remark this answers is not in the document]
Comment (jcapona) [reply further_link/endpoint/upload.py:13]: actually it can be `*.cfg`; however I see the benefit in only looking for `project.cfg` files, we should probably change it to specifically look for that file (here & in the miniscreen app)

File: further_link/util/copy.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+
+async def do_copy_files_to_projects_directory(src_directory, directory, user=None):
+    if user is None:
+        user = default_user()
+    
+    dst_directory = get_miniscreen_projects_directory(
+        directory.get("name"), user, directory.get("username")
+    )
+    rmtree(dst_directory, ignore_errors=True)
```
Comment (angusjfw) [inline further_link/util/copy.py:28]: this will delete the project log file. I think this is what we want, it's best not to have a log file for out of data code, but it might be related to the issues i saw with the logfile when testing. Anyway, I don't think there are any issues here

File: further_link/util/user_config.py [hunk 5]
```diff
@@ -89,6 +91,21 @@ def get_working_directory(user=None):
...
+    if username is None:
+        username = "unknown"
+
+    base_dir = os.path.join(get_home_directory(user), DEFAULT_PROJECTS_DIR)
+
+    from_env = os.environ.get(FURTHER_LINK_MINISCREEN_PROJECTS_DIR)
+    if from_env is not None:
+        base_dir = from_env
+
+    return os.path.join(base_dir, username.split("@")[0], name)
```
Comment (angusjfw) [inline further_link/util/user_config.py:106]: the @ symbol is a bit surprising here without the context. a comment wouldn't hurt

**Review by angusjfw — CHANGES_REQUESTED** [review]
Looks good, but one of the tests you've adapted is broken so I think you should fix that please

**Review by angusjfw — APPROVED** [review]

File: further_link/endpoint/run_py.py [hunk 6]
```diff
@@ -46,6 +46,7 @@ async def handle_message(message, process_handler, socket):
         await do_upload(
             m_data.get("directory"), process_handler.work_dir, m_data.get("user")
         )
+
```
Comment (jcapona) [inline further_link/endpoint/run_py.py:49]:
```suggestion
```
